app/org/sagebionetworks/bridge/services/ScheduledActivityService.java

```diff
@@ -180,13 +180,68 @@ public ScheduledActivityList getActivityHistory(String healthCode,
     }
 
     public List<ScheduledActivity> getScheduledActivities(ScheduleContext context) {
-        checkNotNull(context);
-        return getScheduledActivities(context, V3_FILTER, V3_MERGE);
+        List<ScheduledActivity> scheduledActivities = scheduleActivitiesForPlans(context);
+        List<ScheduledActivity> dbActivities = activityDao.getActivities(context.getEndsOn().getZone(), scheduledActivities);
+        
+        Map<String, ScheduledActivity> dbMap = Maps.uniqueIndex(dbActivities, ScheduledActivity::getGuid);
+        
+        List<ScheduledActivity> saves = performMerge(scheduledActivities, dbMap, V3_MERGE);
+        activityDao.saveActivities(saves);
+        
+        return orderActivities(scheduledActivities, V3_FILTER);
     }
     
     public List<ScheduledActivity> getScheduledActivitiesV4(ScheduleContext context) {
-        checkNotNull(context);
-        return getScheduledActivities(context, V4_FILTER, V4_MERGE);
+        List<ScheduledActivity> scheduledActivities = scheduleActivitiesForPlans(context);
+
+        // Get all persisted activities within the time frame, not just those found by the scheduler (as in v3).
+        Map<String, ScheduledActivity> dbMap = retrieveAllPersistedActivitiesIntoMap(context, scheduledActivities);
+        
+        // Compare scheduled and persisted activities, replacing scheduled with persisted where they exist
+        List<ScheduledActivity> saves = performMerge(scheduledActivities, dbMap, V4_COLLECT_SAVES);
+        activityDao.saveActivities(saves);
+        
+        // We've removed all the persisted activities that were found by the scheduler. Those have been saved
+        // as necessary. The remaining tasks were scheduled based on user-triggered events that can be updated, 
+        // not events like "enrollment" but events like "when this activity is finished." These are now all 
+        // added to the activities that will be returned.
+        scheduledActivities.addAll(dbMap.values());
```

Comment:
Reasoning through this, I'm pretty sure the bug for https://sagebionetworks.jira.com/browse/BRIDGE-1906 is still present here. We save the new tasks into DDB, but then we immediately overwrite the return values with the old DDB values, at least for one getScheduledActivities() call.

This also tells me that the unit tests need to go up a level and call getScheduledActivities() as a whole instead of calling its individual parts. These unit tests were fine back when merge logic was entirely encapsulated in updateActivitiesAndCollectSaves() https://github.com/Sage-Bionetworks/BridgePF/blob/9404a3536ac3826c6b6bc70b5f42789d2507129d/app/org/sagebionetworks/bridge/services/ScheduledActivityService.java#L132 However, now we have merge logic that exists outside of performMerge() that's no longer being tested.

Ultimately, I think what we actually want is just the logic in V3_MERGE. If there's no DDB Activity, or the DDB Activity is in SCHEDULED or AVAILABLE, we should use the newly generated activity. Otherwise, use the one from DDB. We already know this logic works and has been thoroughly tested. The only difference is that in V4, we want to also save expired tasks. It should be fine to extend this logic to V3, since V3 always hides expired tasks anyway.